### ISSUE TITLE
perf: avoid small performance regressions

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#106661]
**Is your feature request related to a problem? Please describe.**

We currently rely on changepoint detection and manual inspection on https://roachperf.crdb.dev/ to catch performance regressions. We are usually only able to claw back large regressions, which are bisected to a commit in a labor-intensive process (which we've been automating a bit, but still).

But looking at many graphs (like kv95) we see that we are continually backsliding. It's easy to see why: allocations are expensive, Go makes it hard to avoid accidental allocations, there is no forcing function in CI, et cetera.

**Describe the solution you'd like**

I'd like some way to avoid this continuous backsliding. My expectation is that the best way to achieve this is by preventing at least some classes of regressions in CI. For example, by comparing op/s on a small set of representative benchmarks against the parent commit, or by controlling for allocations (though regressions can also be introduced by just adding code!)

**Describe alternatives you've considered**

**Additional context**

Jira issue: CRDB-29662

**Comment by nvanbenschoten** [comment]
I've been thinking through how something like this could work recently as well. One suggestion is that we may get better mileage for fine-grained regression detection from larger microbenchmarks than from roachtests. A good microbenchmark to look at is `BenchmarkKV` suite in `pkg/sql/tests`. The suite has SQL and raw KV interface variants. We would prevent a class of regressions by running these benchmarks in CI and asserting on whether any or all of `ns/op`, `B/op`, and `allocs/op` regress.

As Tobi mentioned, we would need a way to signal to CI that a certain regression is expected and permitted for a given PR.

**Comment by srosenberg** [comment]
@herkolategan FYI. If the above checks out, maybe we could surface results of `BenchmarkKV` more prominently, as well as as fail PRs which exhibit large regressions.

**Comment by tbg** [comment]
Reopened since we haven't solved this issue yet.

In light of the spirited discussion on https://github.com/cockroachdb/cockroach/pull/135899 I'd like to propose that we slightly change the scope of the issue away from "avoiding" small perf impressions by blocking merge jobs to providing high-quality **signal** about perf regressions during the PR workflow for a small (initially tiny) set of core benchmarks.

This should check the following boxes:

- sensitivity and specificity: should reliably detect regressions and improvements, avoiding false positives and negatives. (These goals are in tension with each other, but picking highly stable benchmarks and running interleaved for enough iterations helps a lot).
- visibility: information should reach the author proactively but should also be available for those revisiting the PR after.
- actionability: report contains easy-to-use reproduction instructions as well as artifacts that can obviate need to reproduce in many cases (CPU and alloc profiles for example, including delta profiles for before/after the change and instructions on how to view them).
- timely: gathering the information must not delay CI becoming green. It should also operate on PR push CI, not merge CI, as before the merge is the best time to address issues.
- notification hygiene: avoid spamming comment threads etc on repeat pushes to a PR, acknowledging that many PRs don't have a performance component. When there is no meaningful change, there should be no ping, just a green job on the CI matrix that the user can check if they'd like to.

I think we can find 1-3 good initial benchmarks, sourced from `BenchmarkKV`, `BenchmarkSysbench`, and perhaps we can also look at one that more meaningfully exercises SQL/DistSQL (cc @mgartner you seem fond of `BenchmarkEndToEnd`; anything else come to mind?). I'd like to not get hung up on the benchmark selection, as that can vary over time and we could entertain all kinds of heuristics that select the set dynamically to avoid the checks becoming too slow. The goal should be that the benchmarks, including build, finish in 10-15m.

So assume we have good coverage for an initial set of important hot code paths with benchmarks that are fairly stable. We can run them benchdiff-style (say ten reps of old and new binary each, prescribed benchtime-count, run interleaved [old new old new. ....]) and do the benchstat-style regression check. Allocation bytes/counts tend to be more stable than CPU time (in my experience), so we can set different thresholds for each.

If we find a regression (or we could do this unconditionally, too - doesn't take much time) we would also invoke the old and new benchmark an additional time and collect memory and CPU profiles. The report would then contain these alongside the invocations to re-run the comparison locally, as well as the right pprof invocations to view the before, the after, and the delta profiles (after minus before). For starters, a `zip` file with the right files plus a link to a confluence page for pprof commands should do the trick.

I'm a little unsure about notifications. I think a reasonable way to start is that we introduce this check without any notification on the PR; you'll have to click "Details" on the microbench-CI job to see what transpired. Once we've stabilized the check and are confident that it provides good signal, we can move to posting a (helpful) comment on the PR thread should anything have become measurably faster or slower.

Everything past this line is not something I'd consider in the initial scope of this issue.

In principle I would prefer (at some point) a scheme where we don't make multiple such comments but update the existing one. I could even see us inserting this information into the PR message comment itself (but that's likely a great thing to disagree with at least someone on). Adding a label to the PR could be another subtle way to "notify" that there's something of interest.

There are many next steps we can take after this. We can discuss making the checks required (perhaps validating this decision with additional benchmarking if the initial check detects a regression), perhaps re-framing what this means: rather than enforcing that "regressions aren't allowed", we would require the author to acknowledge the regression and provide a justification for why it's okay before proceeding to merge, meaning that we have a paper trail of how the author/reviewers thought about the regression. We can also discuss how regressions should be treated in bors (my hope would be that we just don't need that ever).

**Comment by mgartner** [comment]
> cc @mgartner you seem fond of BenchmarkEndToEnd; anything else come to mind?

I've mainly been using `BenchmarkEndToEnd/kv*` benchmarks to highlight perf gains/losses in common overhead shared by all queries. The queries are very simple, so overhead of checking the plan cache, converting a logical plan to a physical plan, setting up execution infrastructure, etc. stands out—it isn't hidden by complex optimization or execution that could dominate a benchmark.

I've also been using it simply because of my familiarity with it.

I think that `BenchmarkSysbench` should be similar to `BenchmarkEndToEnd` and we probably don't need the latter.

**Comment by rickystewart** [comment]
> Adding a label to the PR could be another subtle way to "notify" that there's something of interest.

Just a thought, `bors` can avoid merging PR's with certain labels. So the job can add a `regression-detected` label that would block `bors` merging, but in the event that you've determined the regression is unimportant/acceptable/etc., you can remove the label. It's an extra bit of bookkeeping to add to the "paper trail" concept you're discussing. I like that this is a manual action that you must deliberately take.

**Comment by dt** [comment]
Just noting a false positive here: https://github.com/cockroachdb/cockroach/actions/runs/12237982508/job/34134987461?pr=137014

**Comment by dt** [comment]
Another false-positive: https://github.com/cockroachdb/cockroach/actions/runs/12242397467/job/34149629585?pr=137028

**Comment by tbg** [comment]
Here's an example of the kind of thing our tooling could point out and discourage the merge of: https://github.com/cockroachdb/cockroach/pull/137542

**Comment by srosenberg** [comment]
> Here's an example of the kind of thing our tooling could point out and discourage the merge of: #137542

Indeed, that was a fairly large regression on microbenches. Confirming that it was reflected in the weekly run. The regressed commit just happens to be the latest on master at the time of the weekly run; we'd obviously need to bisect to find the first commit.

<img width="1728" alt="weekly_microbenches_ci" src="https://github.com/user-attachments/assets/39b79818-281e-4598-9930-90192378a0ba" />

**Comment by tbg** [comment]
@sambhav-jain-16 thank you - much appreciated. c4-standard-8 with the options you suggested sound good. A related talk is "Performance Matters" by Emery Berger and in particular starting [here](https://www.youtube.com/watch?v=r-TLSBdHe1A). Address space randomization is pointed out as another source of variance, though I think in our case of wanting to compare performance between two different versions of the code we should probably keep ASLR on, to avoid more cases where changes that are strictly outside of the purview of the engineers here to control. But either way, great talk.

Dumping some more ideas (for someday):

_Maybe we should enable CPU+mem profiling throughout?_

The benchmark will be somewhat slower, of course, but the point is not to find out the exact speed of the benchmark, but to pinpoint regressions. I've seen some cases of performance changing meaningfully between commits (highly significant via t-test) but these changes ultimately disappearing after a full rebuild the next day[^1]. When something like this occurs, a profile from just one extra run of each is not as easy to use, since the variance isn't smoothed out. If we collected and merged the profiles across all runs of the version, the profile would "resemble" exactly the runs that we're trying to understand the perf characteristics of.

[^1]: https://github.com/cockroachdb/cockroach/pull/139339

_Highlighting allocs_

I would like us to be able to not slip in any allocations on the hot paths (as determined by some set of benchmarks) accidentally, which requires making them visible. `BenchmarkSysbench`'s `oltp_read_write` txn is ~10 statements, so there are lots of allocations per op. If we added an allocation that is there in each statement, we should see it ten times in each statement, i.e. allocs/op should go up by 10. The problem is that we're also running a full cluster whose allocations will also be tracked, and we need to get the test to a stable enough place where a delta of ten can confidently be blamed on there being a new allocation. More so, ideally we point out the functions for which the allocations have changed.

To this end I've played with https://github.com/felixge/sqlprof/ which can load pprof files into a duckdb instance. Allocations in this benchmark should fall into two classes: stuff that happens "only a few times", or at least does not reliably happen with every request, and the "hot path allocations". We should be able to high-pass the combined heap profile to distinguish the two cases. For example, if we run `oltp_read_write` 1000x, and each has ten statements, then any allocation that shows up fewer than 10k times can't be in the hot path. We can discard these, and focus on the rest, which we aggregate by function. We can then, in a suitable format, highlight the functions that allocate more or less per op as before.

Allocation profiles are sampled, so this may still not be stable enough, but it would be good information to overlay over the source code diff nonetheless. For instance, we could generate coverage maps that point out all locations that allocate, which I believe reviewable can display.

I think it would take quite a bit of work to get this right, but I think there's something there.
